src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/SpellCheck/SpellCheckService.cs

```diff
@@ -0,0 +1,137 @@
+            {
+                // Attribute names are text literals, but we don't want to spell check them because either C# will,
+                // whether they're component attributes based on property names, or they come from tag helper attribute
+                // parameters as strings, or they're Html attributes which are not necessarily expected to be real words.
+                if (node.Parent is MarkupTagHelperAttributeSyntax or
+                    MarkupAttributeBlockSyntax or
+                    MarkupMinimizedAttributeBlockSyntax or
+                    MarkupTagHelperDirectiveAttributeSyntax or
+                    MarkupMinimizedTagHelperAttributeSyntax or
+                    MarkupMinimizedTagHelperDirectiveAttributeSyntax or
+                    MarkupMiscAttributeContentSyntax)
+                {
+                    continue;
+                }
+
+                // Text literals appear everywhere in Razor to hold newlines and indentation, so its worth saving the tokens
+                if (textLiteralSyntax.ContainsOnlyWhitespace())
+                {
+                    continue;
+                }
+
+                if (textLiteralSyntax.Span.Length == 0)
+                {
+                    continue;
+                }
+
+                ranges.Add(new((int)VSInternalSpellCheckableRangeKind.String, textLiteralSyntax.SpanStart, textLiteralSyntax.Span.Length));
+            }
+        }
+    }
+
+    private void AddCSharpSpellCheckRanges(ref PooledArrayBuilder<SpellCheckRange> ranges, ImmutableArray<SpellCheckRange> csharpRanges, RazorCodeDocument codeDocument)
+    {
+        var csharpDocument = codeDocument.GetCSharpDocument();
+
+        foreach (var report in csharpRanges)
+        {
+            var absoluteCSharpStartIndex = report.AbsoluteStartIndex;
+            var length = report.Length;
+
+            // We need to map the start index to produce results, and we validate that we can map the end index so we don't have
+            // squiggles that go from C# into Razor/Html.
+            if (_documentMappingService.TryMapToHostDocumentPosition(csharpDocument, absoluteCSharpStartIndex, out var _1, out var hostDocumentIndex) &&
+                _documentMappingService.TryMapToHostDocumentPosition(csharpDocument, absoluteCSharpStartIndex + length, out var _2, out var _3))
+            {
+                ranges.Add(new(report.Kind, hostDocumentIndex, length));
+            }
+        }
+    }
+
+    private static int[] ConvertSpellCheckRangesToIntTriples(ImmutableArray<SpellCheckRange> ranges)
+    {
+        // Important to sort first as we're calculating relative indexes
+        ranges = ranges.OrderAsArray(CompareSpellCheckRanges);
+
+        using var data = new PooledArrayBuilder<int>(ranges.Length * 3);
+
+        var lastAbsoluteEndIndex = 0;
+        foreach (var range in ranges)
+        {
+            if (range.Length == 0)
+            {
+                continue;
+            }
```

Comment:
Is this a common scenario? When does it happen?

Reply:
Yes, the Razor tree is weird.

Given:
```
@functions
{
    void M()
    {
        // Eat more chickin
    }
}
```

There is a 0-length `MarkupTextLiteral` before the `@functions` and after the final close brace